### DISCUSSION TITLE
raspberry 3 - pulsar-client python

### DISCUSSION THREAD

**Discussion by anand-ftc** [apache/pulsar#18423] [Q&A]
I have tried to compile Python client for Raspberry 3 serveral times. But I have ended up not finding the correct dependencies. Can anyone provide me with the wheel file. I tried building with almost all blogs out there

thanks in advance

**Comment by tisonkun** [comment]
Hi @anand-ftc! How did you try to compile it? With current information, I'd say `pip3 install pulsar-client` should work.

cc @BewareMyPower

**Reply by anand-ftc** [reply]
@tisonkun I tried first compiling the cpp shared objects following https://github.com/apache/pulsar-client-cpp#compilation


Then followed the https://github.com/apache/pulsar-client-python to build the python client wheel files. this is where there are errors in finding the dependencies. 

> pi@raspberry:~/sources/pulsar/pulsar-client-cpp $ cmake -DCMAKE_TOOLCHAIN_FILE=./toolchain-arm32.cmake -DCMAKE_FIND_ROOT_PATH=/usr/arm-linux-gnueabihf  .
> -- ARCHITECTURE: x86_64
> -- BUILD_DYNAMIC_LIB:  ON
> -- BUILD_STATIC_LIB:  ON
> -- BUILD_TESTS:  OFF
> -- BUILD_PYTHON_WRAPPER:  ON
> -- BUILD_WIRESHARK:  OFF
> -- BUILD_PERF_TOOLS:  OFF
> -- LINK_STATIC:  OFF
> -- USE_LOG4CXX:  OFF
> -- CMAKE_BUILD_TYPE:  RelWithDebInfo
> CMake Error at /usr/share/cmake-3.7/Modules/FindPackageHandleStandardArgs.cmake:138 (message):
>   Could NOT find Threads (missing: Threads_FOUND)
> Call Stack (most recent call first):
>   /usr/share/cmake-3.7/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
>   /usr/share/cmake-3.7/Modules/FindThreads.cmake:212 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
>   CMakeLists.txt:90 (find_package)
> 
> 
> -- Configuring incomplete, errors occurred!
> See also "/home/pi/sources/pulsar/pulsar-client-cpp/CMakeFiles/CMakeOutput.log".
> See also "/home/pi/sources/pulsar/pulsar-client-cpp/CMakeFiles/CMakeError.log".
> CMake Error: Unable to open check cache file for write. /home/pi/sources/pulsar/pulsar-client-cpp/CMakeFiles/cmake.check_cache
> 

@tisonkun I tried the pip3 install pulsar-client. It says the following

>  Could not find a version that satisfies the requirement pulsar-client (from versions: )
> No matching distribution found for pulsar-client

**Reply by merlimat** [reply]
@anand-ftc Which version of Python? 

Can you post the result of `pip3 debug --verbose`?

**Reply by merlimat** [reply]
Oh, it's ARM32. We only have binaries for ARM64 so far.

**Reply by anand-ftc** [reply]
@merlimat  pip3 debug returns error : unknown command "debug"

@merlimat yes you are correct .. its ARM32

**Reply by yaalsn** [reply]
@anand-ftc Or you can switch to raspberry pi 4B, it's ARM64.

**Reply by Demogorgon314** [reply]
@anand-ftc  The raspberry 3 supports 64-bit OS, See https://www.raspberrypi.com/software/operating-systems/#raspberry-pi-os-64-bit , can you switch the system to 64-bit?

**Reply by merlimat** [reply]
@anand-ftc There is no support for arm32 (`armv7l`) in PyPI and manylinux, so it would make it hard to distribute these wheel files.

**Reply by anand-ftc** [reply]
@yaalsn We have several rpi3 out in the field in several places and we cannot upgrade them now. Thanks

@Demogorgon314  Our rpi runs softwares from multiple teams and changing the OS at this point might be little tricky. I will think of this option and take it up with others.. thanks

**Reply by anand-ftc** [reply]
> @anand-ftc There is no support for arm32 (`armv7l`) in PyPI and manylinux, so it would make it hard to distribute these wheel files.

@merlimat Is it possible to help build a arm32 wheel files so that we could manually deploy and install them.. thanks anyway
